app/controllers/admin/backups_controller.rb

```diff
@@ -173,6 +186,21 @@ def upload_backup_chunk
     render body: nil
   end
 
+  def create_upload_url
+    params.require(:filename)
+    filename = params.fetch(:filename)
+
+    return render_error("backup.backup_file_should_be_tar_gz") unless valid_extension?(filename)
+    return render_error("backup.invalid_filename") unless valid_filename?(filename)
+
+    store = BackupRestore::BackupStore.create
+    upload_url = store.generate_upload_url(filename)
+  rescue BackupRestore::BackupStore::BackupFileExists
```

Comment:
Error handling is really slow in Ruby so we want to avoid using this pattern for control flow. Perhaps we can try using `catch..throw` here?